StreamChatClient_v3/Controllers/ChannelController.swift

```diff
@@ -5,21 +5,60 @@
+    ///
+    /// - Parameter channelQuery: The ChannelQuery this controller represents
+    /// - Returns: A new instance of `ChannelController`.
+    ///
+    public func channelController(for channelQuery: ChannelQuery<ExtraData>) -> ChannelControllerGeneric<ExtraData> {
+        .init(channelQuery: channelQuery, client: self)
+    }
+}
+
+/// A convenience typealias for `ChannelControllerGeneric` with `DefaultDataTypes`
+public typealias ChannelController = ChannelControllerGeneric<DefaultDataTypes>
+
 /// `ChannelController` allows observing and mutating the controlled channel.
 ///
 ///  ... you can do this and that
 ///
-public class ChannelController<ExtraData: ExtraDataTypes>: Controller {
-    public let channelId: ChannelId
-    public weak var delegate: ChannelControllerDelegate?
+public class ChannelControllerGeneric<ExtraData: ExtraDataTypes>: Controller, DelegateCallbable {
+    /// The ChannelQuery this controller observes.
+    public let channelQuery: ChannelQuery<ExtraData>
     
-    private let viewContext: NSManagedObjectContext
-    private let worker: ChannelUpdater<ExtraData>
+    /// The `ChatClient` instance this controller belongs to.
+    public let client: Client<ExtraData>
+    
+    /// The channel matching the channelId. To observe updates to the channel, set your class as a delegate of this controller and call `startUpdating`.
+    public private(set) lazy var channel: ChannelModel<ExtraData>? = {
+        log.warning("Accessing `channels` before calling `startUpdating()` always results in nil channel.")
+        return nil
+    }()
     
-    private lazy var fetchResultsController: NSFetchedResultsController<ChannelDTO> = {
-        let request = ChannelDTO.fetchRequest(for: channelId)
+    /// The worker used to fetch the remote data and communicate with servers.
+    private lazy var worker: ChannelUpdater<ExtraData> = self.environment.channelUpdaterBuilder(client.databaseContainer,
+                                                                                                client.webSocketClient,
+                                                                                                client.apiClient)
+    
+    /// A type-erased delegate.
+    private(set) var anyDelegate: AnyChannelControllerDelegate<ExtraData>?
+    
+    /// Used for observing the database for changes.
+    private(set) lazy var fetchResultsController: NSFetchedResultsController<ChannelDTO> = {
+        let request = ChannelDTO.fetchRequest(for: channelQuery.cid)
         let frc = NSFetchedResultsController(fetchRequest: request,
-                                             managedObjectContext: viewContext,
+                                             managedObjectContext: client.databaseContainer.viewContext,
                                              sectionNameKeyPath: nil,
                                              cacheName: nil)
         frc.delegate = self.changeAggregator
```

Comment:
We will need multiple FRCs here. One for observing the changes in the channel itself, other for observing the list of messages.